Bug fixes + local Watch Later and History lists

MoojMidge opened the pull request:
### Fixes:
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/115
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/250
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/411
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/480
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/530
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/532
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/534
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/536 and hopefully prevents it from reoccurring
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/538
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/540
- Fix https://github.com/anxdpanic/plugin.video.youtube/issues/545
- Possibly fix https://github.com/anxdpanic/plugin.video.youtube/issues/425
- Possibly fix https://github.com/anxdpanic/plugin.video.youtube/issues/434
- Possibly fix https://github.com/anxdpanic/plugin.video.youtube/issues/485
- Workaround crashes in https://github.com/anxdpanic/plugin.video.youtube/issues/540, https://github.com/anxdpanic/plugin.video.youtube/issues/113
- Workaround for https://github.com/anxdpanic/plugin.video.youtube/issues/537

### New:
- Resolves https://github.com/anxdpanic/plugin.video.youtube/issues/18
- Resolves https://github.com/anxdpanic/plugin.video.youtube/issues/450
- Resolves https://github.com/anxdpanic/plugin.video.youtube/issues/464
- Resolves https://github.com/anxdpanic/plugin.video.youtube/issues/502
- Resolves https://github.com/anxdpanic/plugin.video.youtube/issues/503
- Resolves https://github.com/anxdpanic/plugin.video.youtube/issues/539
- Partially resolves https://github.com/anxdpanic/plugin.video.youtube/issues/505
- Adds a local Watch Later and History list for use when not logged in or custom playlist id is not set

MoojMidge committed:
Fix error with subtitle selection prompt

MoojMidge committed:
Misc minor fixes

MoojMidge committed:
Updates to allow backports for Kodi 18

MoojMidge committed:
Update client details

MoojMidge committed:
Fix stream quality comparison to quality selections

Iterating over dict in insertion order only available in Python 3.7+

MoojMidge committed:
Fix inputstream.adaptive dependency for matrix.unofficial

MoojMidge committed:
Test workaround for some MPD live stream manifests

MoojMidge committed:
Debug test

MoojMidge committed:
Decrease inputstream.adaptive dependency version

MoojMidge committed:
Update sort methods

- Remove use of eval
- Remove incrementing dummy values for missing sort methods
- Not sure why it was done this way originally (for old unit tests?)
  but the incrementing dummy values can be added back if needed

MoojMidge committed:
Misc tidy up

No (intended) functional changes:
- extraneous indentation/whitespace
- missing explicit return values
- unnecessary return values
- combine nested if statements
- else/elif after return
- change else if to elif
- change if X: to if not X: return
- use conditional expressions
- simplify redundant conditions in if statements
- change bool if condition else not bool to just condition
- remove u-strings
- remove unused variables
- reduce line lengths

MoojMidge committed:
Tidy up imports

- Group multiple imports
- Remove star imports
- Limit line length
- Use from import where appropriate

MoojMidge committed:
Misc optimisations and refactoring

Needs testing for regressions

MoojMidge committed:
Update release names

MoojMidge committed:
ISA - fix logic & naming for text, vars and methods

- IA to ISA, ia to isa, adaptive to isa
- MPD to ISA, mpd to isa when non MPEG-DASH specific
- new HLS/hls variables and methods for HLS specific functionality

MoojMidge committed:
Add preliminary support for iOS premium streams

Can be enabled from
Settings > Advanced > Use alternate client details > Alternate #1

- Partially fixes #505

MoojMidge committed:
Update defaults as ISA is now a required dependency

MoojMidge committed:
Move youtube requests to login_client

- Use common method for all Youtube requests
- Use common connection pool
- Use as context manager
- TODO migrate subtitle and video_info client requests

MoojMidge committed:
Use new client request context manager

MoojMidge committed:
Ensure headers are added to cURL playback requests

MoojMidge committed:
Split LoginClient into base & client request class

Replace client build and request methods from VideoInfo

MoojMidge committed:
Use super() rather than hardcoded parent class

MoojMidge committed:
Locator as sub-class of BaseRequestsClass

MoojMidge committed:
Re-org to prevent issues with circular imports

- Move http_server, ip_api, and requests to kodion.network module from
  kodion.utils
- Split kodion.impl to
  - kodion.context
  - kodion.player
  - kodion.plugin
  - kodion.settings
  - kodion.ui
- Remove context parameter from BaseRequestsClass, requests now imports
  Settings directly and can be imported by other modules in kodion
- Update BaseRequestsClass.request to use same params as parent class

MoojMidge committed:
Fix completed live streams playing at max 720p

- Fix #530
- Partially fix #540
- MPEG-DASH and HLS manifests are not provided for completed live streams

MoojMidge committed:
Add setting methods to get/set API details

- Also add some basic validation

MoojMidge committed:
Fix local resuming not working after d873bbd

- Also enable for non MPEG-DASH videos
- Also respect user choice from Kodi resume prompt

MoojMidge committed:
Use correct types for context params and play_data

MoojMidge committed:
Add lisitems to directory as complete listing

MoojMidge committed:
Update sqlite based modules

- Faster and maybe more rebust
- Deduplicate code in subclasses, move to Storage
- Fix issue where timestamp was stored as datetime string

MoojMidge committed:
Update listitem creation to fix date sorting

- Fixes #411
- Workaround issue with infotagger until updated
- Fix resume properties
- Update ISA properties

MoojMidge committed:
Update stream selection logic for codecs and HDR

- Fix #532
- Workaround for ISA not initialising decoder when codec changes within Adaptationset
- AV1 and H264 are no longer grouped together in the same AdaptationSet
- HDR is grouped separately for ISA stream selection dialog

MoojMidge committed:
Improve robustness of JSONStore

- Fix #536 and hopefully prevents it from reoccuring

MoojMidge committed:
Update date processing and display

- Remove workarounds for total_seconds()
- Fix issues with incorrect date display
- Fix issue with storing incorrect date values
- Fix mixups with dateadded, date, premiered, etc.
- Fixes #464
- Possibly fixes #425, #434

MoojMidge committed:
Update label formatting methods

- Adds new format:
  - [LIGHT]blah[/LIGHT]
  - [I]blah[/I]
  - [TABS]x[/TABS]blah
  - [CR]blah and/or blah[CR]
- Optional [CR] to other format methods

MoojMidge committed:
Retrieve and display video stats

- Existing "Show channel name in description" option changed to
  "Show channel name and video details in description"
- Displays number of views, likes, comments in description if enabled
- Displays stats and premiere datetime in label2
- Closes #503, #18

MoojMidge committed:
Remove duplicated update_video_infos code

MoojMidge committed:
Update UrlToItemConverter for timestamps

- Also support live videos and playlists
- Partially fix #115
- Fix #502
- Fix #538
- Remove unnecessary seek on start when resuming

MoojMidge committed:
Consolidate localisation ids

- Move LOCAL_MAP from provider to context as that is where it is used
- Remove const_localize (redundant when using LOCAL_MAP)
- Remove _local_map (redundant when using LOCAL_MAP)
- Update and sort id names
- Replace hardcoded integer ids with id names
- Tidy up provider

MoojMidge committed:
Remove references to xbmc.translatePath

- Also remove unnecessary decode to unicode
- Future backports for Kodi 18 to use six/future

MoojMidge committed:
Update make_comment_item

- Use ui label formatting methods
- Fix date and dateadded

MoojMidge committed:
Update provider._on_users

- Remove duplicated code
- Change to int user indexing rather than integers cast as str
- Fix issues with incorrect sorting of user (dict retaining insertion ordering requires Python 3.7+)

MoojMidge committed:
Fix issues with (non)playable items

- Fix UriItem always using setResolvedUril rather than RunPlugin
- Ensure ListItem is created with path
- Standardise naming of ListItem creation methods
- Shorten name of (get|set|clear)_home_window_property to (get|set|clear)_property
  - Use string type return rather than None
- Fix setting isPlayable based on item type
- Fix filtering shorts based on duration of playable items
- Remove filtering shorts on list of non playable items
- Fix routing of /play/ with playlist_id/playlist_ids params
- Fixes #115 and #480
- Possibly fixes #485

MoojMidge committed:
Misc tidy up

- No (intended) functional changes

MoojMidge committed:
Fix missing utils.to_str in __all__

MoojMidge committed:
Minor improvement to update_video_infos

- Fix watch later context menu item on items already in WL
- Update regexps
- Minimise work in loop

MoojMidge committed:
Update video stats display

- Uses same formatting style as comment details

MoojMidge committed:
Update subtitles to use BaseRequestClass

MoojMidge committed:
Update url_resolver to use BaseRequestClass

- Update user-agent
- Use retry and error handling functionality of BaseRequestClass to fix #250
- Simplify/update resolver logic for new URLs and meta content
- Needs testing for regressions

MoojMidge committed:
Improve UrlToItemConverter channels and playlists functionality

- Add support for live channels
- Add support for playlist starting video_id
- Simplify regexp

MoojMidge committed:
Add option to disable HFR video at max resolution

- Closes #539

MoojMidge committed:
Improve video stats display

- Fix incorrect duration of upcoming videos
- Fix imdbnumber being saved to incorrect info label
- Add view count to listitem info and sort methods
- Add rating to listitem info and sort method based on made up ranking
- Add label masking to sort methods

MoojMidge committed:
Fix incorrect usage of DataCache in VideoInfo

- Move class variable to Storage and access via instances
- Change parameter order for DataCache.getItem[s] and FunctionCache.get
  - first param is now item/func to
  - second param is now time in seconds
- Optimise imports

MoojMidge committed:
Rationalise item date and duration methods

- use datetime objects for setting rather than parsing back and forth
- date and dateadded values stored as datetime rather than str
- date and dateadded getter can retrieve value as datetime, date str or datetime.isoformat
- aired and premiered values stored as date rather than st
- aired and premiered getter can retrieve value as date, or date.isoformat
- datetime_parser.parse will always create a datetime object rather than a date object
  - the caller can determine whether datetime.date() needs to be used after that
- duration getter can optionally retrieve value as hh:mm:ss string
- listitem label2 now uses the same data as default label mask including date and duration

MoojMidge committed:
Add support for start and end parameters

MoojMidge committed:
Update live status based on current video details

- Partially fix #540
- Also properly select protocol type for InputStream Helper

MoojMidge committed:
Support Youtube Clips

- Closes #450
- Support following plugin url formats:
  - plugin://plugin.video.youtube/uri2addon/?uri=https://www.youtube.com/clip/<clip_id>
  - plugin://plugin.video.youtube/play/?video_id=<video_id>&start=<start_time>&end=<end_time>&clip=true
- optional clip parameter will loop the clip between start_time and end_time, otherwise video will stop after end_time is reached

MoojMidge committed:
Workaround Kodi crash due to multiple busy dialogs

- Workaround crashes in #540, #113
- Won't fix all busy dialog crashes due to how Kodi resolves plugin urls, but should fix most common crashes

MoojMidge committed:
Tidy up and reformat

- store context as _context instance variable
- change YoutubePlayer time variables to have non-protected names
  - seek_time, start_time, end_time
- handle watch history logging in update_watch_history

MoojMidge committed:
Add ForceResolvePlugin property to listitem

- Attempt to reduce edge cases that can Kodi to crash due to multiple busy dialogs

MoojMidge committed:
Update monitor and http_server

- Follow up to 6d1301ea
- Fix for server settings not applying until addon restarts
- Use BaseRequestsClass
- Update formatting
- Use Settings rather than getting addon settings directly
  - this is in preparation for using new settings class via xbmcaddon.Addon('id').getSettings()
  - aim to update settings on onSettingsChanged rather than reading and parsing settings xml all the time
- Add basic validation of port setting

MoojMidge committed:
Fix regressions in live channels following 72480b5

MoojMidge committed:
Remove duplicate code in VideoInfo._get_error_details

MoojMidge committed:
Update LoginClient and YouTubeRequestClient

- Update formatting
- Use LoginException as default exception type for LoginClient
- Fix for possible undefined instance variables
- TODO: use build_client for LoginClient

MoojMidge committed:
Updates to listing display

- Remove unused date/time formatting code
- Use locale appropriate date/time display rather than ISO format
- Standardise stats item separator to match label mask
- Don't bother including all data in label2 as it just duplicates details in skins that use it
- Fix non-upcoming videos showing date/time in italics
- Fix showing zero value stats
- Fix stats with a value of 1 referred to as plural/multiple
- Fix incorrect date sorting after e16c7a0 (Kodi datetime need 'T' as seperator)

MoojMidge committed:
Disable Opus audio by default

- Workaround for #537

MoojMidge committed:
Remove debugging code

MoojMidge committed:
Kodi Leia compatibility

MoojMidge committed:
Fix incorrect method signatures

MoojMidge committed:
Fix datetime_parser.now being removed

- Still used in YouTube._get_recommendations_for_home
- TODO: remove once _get_recommendations_for_home is fixed

MoojMidge committed:
Add settings methods for WL and HL

MoojMidge committed:
Better handle failed playlist request

- Fix #545
- Also consolidate progress dialog methods

MoojMidge committed:
Better handle seeking out of a Clip

- This may need to be tweaked after more testing

MoojMidge committed:
Make merge_dicts a common utils method

MoojMidge committed:
Make AccessManager inherit from JSONStore directly

- Remove LoginTokenStore and login_tokens.py
- Prevents issues with possible circular imports
  - Move access_manager out of kodion.utils and into kodion.json_store
  - Remove JSONStore from kodion.json_store.__all__
- JSONStore.save can now partially update data
- JSONStore.load/save/get_data can now process data after JSON decoding
- TODO: Fix usage of AccessManager in youtube.client.__config__.APICheck to avoid keeping two copies of the data
- TODO: Fix usage of APIKeyStore to avoid keeping two copies of the data

MoojMidge committed:
Move sql storage based classes to new kodion.sql_store module

- Originally located in kodion.utils
- Prevents possible issues with circular imports

MoojMidge commented:
Some other potential issues with circular imports have been identified. Additional fixes required.

MoojMidge committed:
Use single instance of SystemVersion

MoojMidge committed:
Add get_kodi_setting method

MoojMidge committed:
Update settings to use xbmcaddon.Settings class

- Used in Kodi v20+
- Backwards compatibility also maintained
- Values are cached, flushed when onSettingsChanged callback runs
- Values are logged when debug logging is enabled
- TODO: Use single instance of XbmcPluginSettings

MoojMidge committed:
Make better use of new AccessManager

- Update youtube.client.__config__
- Follow up to f5742c8

MoojMidge committed:
Remove/add Unicode encoding/decoding

- Additions required for Python 2/3 compatibility
- Removals not required for Python 3
  - Will require use of kodi-six in Kodi Leia

MoojMidge committed:
Workaround for missing datetime.timestamp method in Python < v3.3

- Also update default fmt parameter in strptime wrapper

MoojMidge committed:
Add Python2 and Kodi 18 compatibility tweaks

- Ensure new style classes are used
- Use correct inputstream property for listitems
- Use correct infolabel date formatting
- Use correct setting type getter/setter for str list
- Ignore labelmask parameter in sort methods
- Use traditional packages rather than namespace packages

MoojMidge committed:
Ensure current_user value is loaded as int

MoojMidge committed:
Prevent possible memory leak and/or freeze on exit

MoojMidge committed:
Fix incorrect string id

MoojMidge committed:
Remove need for urlparse and urlunsplit usage

- Replace urlparse usage with urlsplit instead
- Replace urlunsplit with _replace().geturl()

MoojMidge committed:
Import tidy up

MoojMidge committed:
Update filter_short_videos

- Caller must check whether Settings.hide_short_videos() is True
- update_video_infos will short circuit if short videos are hidden

MoojMidge committed:
Improve default sorting of live videos

- aired infolabel does not include time, date does
- sort by date/time ascending

MoojMidge committed:
Minor speed improvements to creation of listings

- TODO: Proper thread management
- Reduce requests backoff factor

MoojMidge committed:
Avoid JSONStore.save if data is unchanged on update

- Also update log messages

MoojMidge committed:
Add print_stats convenience method to debug.Profiler

MoojMidge committed:
Fix re-requesting blocked video info

- Cache empty video data
- May result in failed requests being cached for a month
- TODO: See whether this will require changes to cache expiration

MoojMidge committed:
Misc tidy ups

- time_milliseconds parameter for notifications renamed to time_ms
- Updated imports

MoojMidge committed:
Remove unused code

- Introduced as part of reverted tags PR

MoojMidge committed:
Create custom InvalidJSON exception

- Inherits from requests.exceptions.InvalidJSONError and KodionException

MoojMidge committed:
Fixup log out on key change after b3b08d2

MoojMidge committed:
Remove all key set setting related functions

- Fixes after ba686a5 and b3b08d2
- Functionality was deprecated and should have been removed not refactored

MoojMidge committed:
Reduce whitespace in description

- Looks worse but there is limited space on many list/widget views

MoojMidge committed:
Fix not updating channel fanart after 0889e7b

MoojMidge committed:
Update response/error hook implementation

MoojMidge committed:
Standardise input to SQL storage

- Input/output will be always be JSON de(serialized) and (un)pickled
- Also set ensure_ascii=False

MoojMidge committed:
Add context manager to progress dialogs

MoojMidge committed:
Fix incorrect date sorting after a3af85f

- Sorting was messed up due to sorting based on localised date format
- No default sorting for non-live listings, will use order from API
- TODO: Ensure correct order is kept when using cache

MoojMidge committed:
Use same response/error hooks for Youtube.perform_v3_request

- Partially fixes #545
- Remove the other handle_error methods

MoojMidge committed:
Add caching of playlistitems requests

- Partially fixes #545
- Results will be cached for 1 hour
- Also re-sorts cached values so they match the original requested order

MoojMidge committed:
Remove JSON (de)serialisation for SQL storage

- Partially revert 764caec which added missing JSON (de)serialisation
- Data is already pickled and stored as a binary blob
- Add versioning via table name
- Bump to v2 to force cached data to be removed
- Old table(s) will be automatically removed
- Also remove row factory

MoojMidge committed:
Fix initial discard of subsequent pages of playlistitems

MoojMidge committed:
Minor optimisations for SQL storage

- Remove unnecessary type conversions
- Remove unnecessary sorting
- Remove unnecessary iteration

MoojMidge committed:
Update type/attr/key checking

- Allow passing tuples and dict.keys instead of only lists as parameters
- Other misc fixes to prevent possible exceptions

MoojMidge committed:
Remove old data on optimise SQL storage by size

- Also add logging for errors
- Bump storage table version

anxdpanic reviewed (approved):
Please go ahead and merge this when you're ready @MoojMidge 

Also, once merged, submit and merge a beta bump and ping me and I will update the repositories with the beta 👍 

Thanks again for you help with this add-on and much more!

MoojMidge commented:
All good. There are a couple more things I messed up that I will need to update in this, then ready for a beta release

MoojMidge committed:
Misc tidy ups

MoojMidge committed:
Further updates to SQL storage

- Versioning for each type of storage
- Table creation and update to occur once per class or if db is deleted
- Use connection context managers to auto-commit/rollback
  - Transactions created manually where required
- Restore Python 2 compatibility

MoojMidge committed:
Update ResourceManager and _process_list_response

- Defer cache store until all data has been fetched/updated
- Improve deferred thread creation sequence

MoojMidge committed:
Create Session as class variable in BaseRequestsClass

- Create once, close pools when not needed

MoojMidge committed:
Set _is_action in DirectoryItem.__init__

MoojMidge committed:
Add local Watch Later and History lists

- Make items.utils.to_json a method of BaseItem
  - Also fix to allow handling of date/datetime objects
  - Will only work in Python 3.7+
- TODO: fix date parsing for older versions of Python
- TODO: update context menu items
- TODO: fix serialisation prior to updating items

MoojMidge committed:
Initial consolidation of context menu methods

- Context menu items are evaluated at the time that they are created
- This creates problems when the underlying item data changes
- The context menu item can then contain out of date data
- e.g. updated channel fanart or video play data
- TODO: register menu items, but don't call them until ListItem is added to UI

MoojMidge committed:
Provide stack trace in error log messages

MoojMidge committed:
Allow for proper inheritance of request exceptions

- Also tidy up imports after 860defd

MoojMidge committed:
Create new Context without default plugin_id value

MoojMidge committed:
Youtube class now requires Context as a parameter

- Reuse existing, rather than creating a new Context at the module level
- Remove log method callbacks

MoojMidge committed:
Fix regression after 72480b5 caused by typo

MoojMidge committed:
Fix incorrect MPEG-DASH mime type

MoojMidge committed:
Remove redundant is_empty method on Storage sub-classes

- Just use re-named base class method

MoojMidge committed:
Change constants module import aliases

- "content_type" to "content"
- "setting" to "settings"
- "sort_method" to "sort"

MoojMidge committed:
Remove use of hardcoded temp and data path

- make_dirs returns translated path on success
  - Use to check/create/get in one go
- Rename native_path to addon_path
- Use os.path.join rather than hardcoding path seperators
- Fix unnecessarily nested os.path.join calls

MoojMidge committed:
Remove use of hardcoded addon id and path

MoojMidge committed:
Update item creation

- Add resource path constant
  - Removes multiple function calls to set default fanart
- Simplify parameter creation/updates
- Use plugin path constants rather than hardcoded paths
- Use fanart parameter in *Item.__init__
- Use action parameter in DirectoryItem.__init__
- Add new location icon for My Location directory item
- Add new icons with consistent sizing for the various search functions

MoojMidge committed:
Add http server path constants

- Adds "/youtube/" to path
  - for future ISA requirements to identify license provider

MoojMidge committed:
Misc tidy ups

MoojMidge committed:
Update Auto remove from Watch Later for local list

- Rename classes and module
- Misc tidy up
- Prevent potential indefinite lockup on Thread.join()

MoojMidge committed:
Rename YouTubeMonitor to ServiceMonitor

MoojMidge committed:
Test SQL storage REPLACE speed improvements

MoojMidge committed:
Simplify utils.methods.make_dirs add rm_dir

- Also make code Python 2 compatible (remove exist_ok param in os.makedirs)

MoojMidge committed:
Don't add .sqlite file extension when creating db

- Correct filepath to be used when creating new instance
- TODO: add filenames to constants

MoojMidge committed:
Simplify datetime.strptime bug workaround

- Use common method in other modules
- Fix for Python2 compatibility as follow up to 421ec80

MoojMidge committed:
sqlite.BINARY Python2 compatibility fix

- BINARY is memoryview in Python3, buffer in Python2

MoojMidge committed:
Misc tidy ups

MoojMidge committed:
Beta bump - v7.0.3+beta.1

MoojMidge commented:
@anxdpanic - I couldn't easily separate out the changes, so this also includes the second (side) step in restoring video recommendations - a local only history list, used when not logged in, or when the custom history playlist id is not set.

Requires enabling `Settings > Advanced > Use local playback history`

A local Watch Later list has also been added.

Will update the unofficial patches a bit later.
